File: Algorithms/031. Next Permutation/Solution.java
public class Solution {
    public void nextPermutation(int[] nums) {
    	if (nums.length>1) {

        	int a=0;
        	for (int i = 1; i < nums.length; i++) {
    			if (nums[i]>nums[i-1]) {
    				a=i;
    			}
    		}
            for (int i = a; i < (a+nums.length)/2; i++) {
    			int k=nums[i];
    			nums[i]=nums[nums.length-i+a-1];
    			nums[nums.length-i+a-1]=k;
    		}
            if (a>0) {
            	for (int i = a; i < nums.length; i++) {
            		if (nums[a-1]<nums[i]) {
            			int k=nums[i];
            			nums[i]=nums[a-1];
            			nums[a-1]=k;
            			break;
            		}
            	}
    		}
		}
    }
}

File: Algorithms/031. Next Permutation/Solution.py
#coding=utf-8
__author__ = 'xuxuan'
class Solution(object):
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        start=-1
        for i in range(len(nums)-2,-1,-1):
            if nums[i]<nums[i+1]:
               start=i
               break
        if start==-1:
            nums.sort()
        else:
            end=0
            for i in range(start+1,len(nums)):
                if nums[i]>nums[start]:
                    end=i
            arrays=nums[start:end]+nums[end+1:]
            arrays.sort()
            nums[start]=nums[end]
            for i in range(start+1,len(nums)):
                nums[i]=arrays[i-start-1]